benchmarks/cpp/src/coro-qsort.cpp

```diff
@@ -0,0 +1,72 @@
+#include <algorithm>
+#include <array>
+#include <asio.hpp>
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <vector>
+
+using namespace std::chrono;
+using asio::awaitable;
+using asio::co_spawn;
+using asio::detached;
+
+awaitable<void> quickSort(asio::io_context &ctx,
+                          std::vector<int>::iterator begin,
+                          std::vector<int>::iterator end) {
+  if (std::distance(begin, end) <= 32) {
+    // Use std::sort for small inputs
+    std::sort(begin, end);
+  } else {
+    auto pivot = begin + std::distance(begin, end) - 1;
+    auto i = std::partition(begin, pivot, [=](int x){ return x <= *pivot; });
+    std::swap(*i, *pivot);
+
+    auto quickSortWrapper = [&](auto &&begin, auto &&end) {
+      quickSort(ctx, begin, end);
+    };
+
+    co_await quickSort(ctx, begin, i);
+    co_await quickSort(ctx, i + 1, end);
```

Comment:
These parts still need to be done concurrently to each other, otherwise this is a synchronous quickSort(). From [this post](https://cppalliance.org/richard/2021/05/30/RichardsMayUpdate.html) it looks like you can create an awaitable handle to co_spawn, so:
```cpp
auto left = co_spawn(ctx, [&]() -> quickSort(ctx, begin, i), aio::use_awaitable);
auto right = co_spawn(ctx, [&]() -> quickSort(ctx, i + 1, end), aio::use_awaitable);
co_await left;
co_await right;
co_return;
```

Reply:
don't work with latest version 1.24.
Missing awaitable<T>{qsort(...)}.

And by readjusting:
```c++
auto left = co_spawn(ctx, quickSort(ctx, begin, i), asio::use_awaitable);
    auto right = co_spawn(ctx, quickSort(ctx, i + 1, end), asio::use_awaitable);

    co_await left; // Error: call deleted awaitable
    co_await right;
    co_return;
```
**output:**

```bash
error(compilation): clang failed with stderr: /home/kassane/Documentos/Projetos/zap/benchmarks/cpp/src/coro-qsort.cpp:32:14: error: call to deleted constructor of 'awaitable<void, asio::any_io_executor>'
```

Reply:
Works:

```c++
co_await co_spawn(ctx, quickSort(ctx, begin, i), asio::use_awaitable);
co_await co_spawn(ctx, quickSort(ctx, i + 1, end), asio::use_awaitable);

    // co_await left;
    // co_await right;
    co_return;
```
running
took 3024ms

Reply:
Hm, the issue is `co_await`ing the first recursion which makes it a dependency of the second and prohibits running them concurrently. Any scheme that does this is effectively a slower synchronous version. Is there a way to get around this?

Reply:
I did some tests that did not work significantly.
Including the change below to try to stop the counting error from breaking the operation by throwing an exception.
```c++
co_spawn(ctx, quickSort(ctx, begin, i), asio::use_awaitable); // emit exception - no sort array
// co_spawn(ctx, quickSort(ctx, begin, i), asio::use_future); // no error (took 2742ms)
co_await quickSort(ctx, i + 1, end);
```
I also removed the `co_await` from `co_spawn` in main function. Not much difference in the result, except that if you use `use_awaitable` it also throws an exception.

Reply:
use_awaitable is lazy with any operations, so it'll only make progress once you co_await it.

Reply:
@kprotty, with the new parallel example. I still need to readjust about it.

```bash
filling
shuffling
running
took 319ms
```
But the qsort example itself for c++ will include even more overhead.
Perhaps I would need to redefine the example in C++ for all corresponding examples.

Reply:
I found information about what happens in co_spawn.

>"Make sure you never use reference capturing lambdas which are coroutines. The first time they suspend the lambda object gets destroyed. I find myself almost always writing reference capturing lambdas, so I ended up writing some C++ 20 type constraints which stop the compiler from letting me write reference capturing coroutine lambdas (just compare their sizeof to one) and now I'm good." [14ned - Reddit](https://www.reddit.com/r/cpp/comments/u3igrf/comment/i4qdsxw/)